ktor-shared/ktor-shared-events/common/src/io/ktor/events/Events.kt

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2014-2021 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.events
+
+import io.ktor.util.collections.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.internal.*
+
+public class Events {
+    @OptIn(InternalCoroutinesApi::class)
+    private val handlers = CopyOnWriteHashMap<EventDefinition<*>, LockFreeLinkedListHead>()
+
+    /**
+     * Subscribe [handler] to an event specified by [definition]
+     */
+    public fun <T> subscribe(definition: EventDefinition<T>, handler: EventHandler<T>): DisposableHandle {
+        val registration = HandlerRegistration(handler)
+        handlers.computeIfAbsent(definition) { LockFreeLinkedListHead() }.addLast(registration)
+        return registration
+    }
+
+    /**
+     * Unsubscribe [handler] from an event specified by [definition]
+     */
+    public fun <T> unsubscribe(definition: EventDefinition<T>, handler: EventHandler<T>) {
+        handlers[definition]?.forEach<HandlerRegistration> {
+            if (it.handler == handler) it.remove()
+        }
+    }
+
+    /**
+     * Rise an event specified by [definition] with the specified [value] and call all handlers
```

Comment:
Raises the event specified by [definition] with the [value] and calls all handlers.